### PR TITLE
[11.0][purchase_procurement_run_buy_hook]

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
Purchase Procurement Run Buy Hook
=================================

This module adds hooks to method '_run_buy' of model 'procurement.rule'. The
currently supported hooks are:

* Allow to add more data to an existing purchase order line, when a
  procurement is extending it.

### CODE REVIEW

**Comment by rousseldenis** [comment]
@jbeficent You cannot do that as you don't know if other modules inherit the _run_buy function and you are not sure of the order modules are loaded in stack!

Make a PR on odoo/odoo, usually they fast merge such hooks.

**Comment by JordiBForgeFlow** [comment]
@rousseldenis Thanks. I will try now. In the past I was quite unsuccessful.

**Comment by JordiBForgeFlow** [comment]
@rousseldenis Done here: https://github.com/odoo/odoo/pull/21123. Let's pray now...

**Comment by JordiBForgeFlow** [comment]
@rousseldenis Does not look that the PR that I opened to Odoo is going to be addressed soon. So I will continue with this PR. I have done a refactor to improve compatibility of the hook with multiple databases, and respecting inheritance chain.

**Comment by rousseldenis** [comment]
@jbeficent I see, I was too optimistic about such PR's on Odoo...

**Comment by JordiBForgeFlow** [comment]
@rgarnau @lreficent Can you review?

**Review by rgarnau — APPROVED** [review]
LGTM :+1:

File: purchase_procurement_run_buy_hook/hooks.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+from odoo.addons.purchase.models.purchase import ProcurementRule
+
+
+def post_load_hook():
+
+    @api.multi
+    def _new_run_buy(self, product_id, product_qty, product_uom,
+                     location_id, name, origin, values):
+        if not hasattr(self, '_prepare_purchase_order_line_update'):
+            return self.action_invoice_create_original(
```
Comment (LoisRForgeFlow) [inline purchase_procurement_run_buy_hook/hooks.py]: Where is this declared? I am not able to find it

**Review by LoisRForgeFlow — COMMENTED** [review]
Interesting way to add the hook! One comment:

File: purchase_procurement_run_buy_hook/hooks.py (hunk 1 above)
Comment (JordiBForgeFlow) [reply purchase_procurement_run_buy_hook/hooks.py]: Ask @rgarnau

Comment (JordiBForgeFlow) [reply purchase_procurement_run_buy_hook/hooks.py]: @lreficent you're right. I was calling a wrong method. I now changed to call '_original_run_buy'.

**Comment by JordiBForgeFlow** [comment]
@rousseldenis @lreficent can you update your review?

**Review by LoisRForgeFlow — APPROVED** [review]
LGTM now!

**Review by rousseldenis — APPROVED** [review]
It seems ok!

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: This is compatible with modules that need the  hook and modules that don't need the hook. :cool:

**Pull request merged**